### PR TITLE
Fix  MaybeApplyTLSModeLabel function

### PR DESCRIPTION
**Pull request by zengyuxing007**
Fix MaybeApplyTLSModeLabel function which have potential risk of crash.

I've encountered this crash in 1.15 and 1.16 Istio environments, but I don't quite know how to reproduce it

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Review by bleggett — APPROVED** [review]
Normally I don't like fixing symptoms, but given the existing issue/race condition ref (https://github.com/istio/istio/issues/34227), this makes sense as a guard.

**Comment by zengyuxing007** [comment]
> Normally I don't like fixing symptoms, but given the existing issue/race condition ref (#34227), this makes sense as a guard.

ok，I've found the stack information from the previous crash.

```shell
info  ads  [CD](https://github.com/istio/istio/pull/43752#)[S](https://github.com/istio/istio/pull/43752#): PUSH for node:sleep-b-68c8879466-2nqtx.mytest resources:54 size:55.6kB cached:12/50
panic: assignment to entry in nil map

goroutine 854 [running]:
istio.io/istio/pilot/pkg/networking/util.MaybeApplyTLSModeLabel(0xc0019eb8b0, {0xc000e91f98?, 0x5?})
  istio.io/istio/pilot/pkg/networking/util/util.go:492 +0x2aa
istio.io/istio/pilot/pkg/xds.(*EndpointBuilder).buildLocalityLbEndpointsFromShards(0xc002788c78, 0xc00296e210, 0xc000f92d20)
  istio.io/istio/pilot/pkg/xds/endpoint_builder.go:295 +0xbbc
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).llbEndpointAndOptionsForCluster(0x8?, {{0xc001a46770, 0x66}, {0x0, 0x0}, {0x39531b0, 0x59d8e00}, {0xc0007a9e60, 0x21}, 0xc0012aade0, ...})
  istio.io/istio/pilot/pkg/xds/eds.go:308 +0x26f
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).generateEndpoints(0xc001a46770?, {{0xc001a46770, 0x66}, {0x0, 0x0}, {0x39531b0, 0x59d8e00}, {0xc0007a9e60, 0x21}, 0xc0012aade0, ...})
  istio.io/istio/pilot/pkg/xds/eds.go:312 +0x5a
istio.io/istio/pilot/pkg/xds.(*EdsGenerator).buildEndpoints(0xc0001426d8, 0xc000e59020?, 0xc00249baa0, 0xc001350150)
  istio.io/istio/pilot/pkg/xds/eds.go:513 +0x698
istio.io/istio/pilot/pkg/xds.(*EdsGenerator).Generate(0x2df1fe0?, 0xc0013a69c0?, 0xc001cfde00?, 0xc00249baa0?)
  istio.io/istio/pilot/pkg/xds/eds.go:387 +0xa5
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushXds(0xc00137de00, 0xc00143aa00, 0xc001350150, 0xc00249baa0)
  istio.io/istio/pilot/pkg/xds/xdsgen.go:116 +0x432
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushConnection(0xc00137de00, 0xc00143aa00, 0xc00143aa00?)
  istio.io/istio/pilot/pkg/xds/ads.go:756 +0x1a5
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).Stream(0xc00137de00, {0x3972038, 0xc001742ca0})
  istio.io/istio/pilot/pkg/xds/ads.go:340 +0xa70
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).StreamAggregatedResources(0xc0007e8b80?, {0x3972038?, 0xc001742ca0?})
  istio.io/istio/pilot/pkg/xds/ads.go:237 +0x25
github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3._AggregatedDiscoveryService_StreamAggregatedResources_Handler({0x344ce20?, 0xc00137de00}, {0x396dd80?, 0xc0012809c0})
  github.com/envoyproxy/go-control-plane@v0.10.3-0.20220719090109-b024c36d9935/envoy/service/discovery/v3/ads.pb.go:298 +0x9f
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0012d0a80, {0x3974d00, 0xc0007591e0}, 0xc001284240, 0xc001ba0f60, 0x5957ea0, 0x0)
  google.golang.org/grpc@v1.47.0/server.go:1542 +0xf5c
google.golang.org/grpc.(*Server).handleStream(0xc0012d0a80, {0x3974d00, 0xc0007591e0}, 0xc001284240, 0x0)
  google.golang.org/grpc@v1.47.0/server.go:1624 +0x9ea
google.golang.org/grpc.(*Server).serveStreams.func1.2()
  google.golang.org/grpc@v1.47.0/server.go:922 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
  google.golang.org/grpc@v1.47.0/server.go:920 +0x28a
```

**Comment by zengyuxing007** [comment]
@bleggett  Please add ”release-notes-none“ label  for this PR, thanks

**Comment by hzxuzhonghu** [comment]
LGTM, this can occur when 
```
func AppendLbEndpointMetadata(istioMetadata *model.EndpointMetadata, envoyMetadata *core.Metadata,
) {
	if !features.EndpointTelemetryLabel || !features.EnableTelemetryLabel {
		return
	}
```

**Comment by hzxuzhonghu** [comment]
@zengyuxing007 Can you please add a test for this

**Comment by zengyuxing007** [comment]
> 

Done

**Comment by zengyuxing007** [comment]
>
> @zengyuxing007 Can you please add a test for this

@hzxuzhonghu  Thanks for you reply,

I have added a test for this function, please take a look again, thank you.

**Review by GregHanson — APPROVED** [review]
thanks for the additional tests!

File: pilot/pkg/networking/util/util_test.go [hunk 1]
```diff
@@ -1227,3 +1227,269 @@ func TestStatefulSessionFilterConfig(t *testing.T) {
 		})
 	}
 }
+
+func TestEndpointTLSModeLabel(t *testing.T) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/util/util_test.go:1231]: This test does not make much sense, i meant testing on the `buildLocalityLbEndpointsFromShards` is better

**Comment by rohitagarwal003** [comment]
Can we cherry-pick this to 1.17 and 1.16?

**Pull request merged**